### ISSUE TITLE
expect_lint_free() fails when Check is launched from RStudio, but not when launched by devtools::check()

### ISSUE THREAD

**Issue by kloppen** [r-lib/lintr#421]
I've set up a test for a package using `testthat` that calls `lintr::expect_lint_free()`. When running this test through RStudio (ie. ctrl+shift+T), the test passes. When running `devtools::check()`, the test (which runs as part of R CMD check) passes. But, when running Check from RStudio (ie. ctrl+shift+E), the test fails with the following error:

```
> library(testthat)
> library(LintRErrorTest)
> 
> test_check("LintRErrorTest")
── 1. Error: Package Style (@test_lint.R#4)  ────────────────────────────────────────────────────────────
invalid 'path' argument
1: lintr::expect_lint_free() at testthat/test_lint.R:4
2: lint_package(...)
3: normalizePath(path, mustWork = FALSE)
4: path.expand(path)

══ testthat results  ════════════════════════════════════════════════════════════════════════════════════
[ OK: 0 | SKIPPED: 0 | WARNINGS: 0 | FAILED: 1 ]
1. Error: Package Style (@test_lint.R#4) 
```

This can be reproduced as follows:

- Create a new package using RStudio's File -> New Project -> R Package
- Run `usethis::use_testthat()` to start using `testthat`
- Create the file `tests/testthat/test_lint.R` with the following content:

```r
if (requireNamespace("lintr", quietly = TRUE)) {
  context("linting package")
  test_that("Package Style", {
    lintr::expect_lint_free()
  })
}
```

- Run tests using either ctrl+shift+T or `devtools::test()`. There will be a test failure caused by a trailing whitespace in a comment in the default contents of `hello.R`. Remove this trailing whitespace if desired.
- Run `devtools::check()`. This will give a warning about a non-standard license, and a test failure caused by the trailing whitespace if you didn't fix it above.
- Run Check from the RStudio IDE (ctrl+shift+E). This will fail with the `invalid 'path' argument` error above.

I have tested this using R 3.6.1 running on Ubuntu 18.04.3 LTS. I currently have `testthat` version 2.2.1 installed. The same behavior occurs with both `lintr` 2.0.0 from CRAN and version 2.0.0.9000 installed using `devtools::install_github`

**Comment by kloppen** [comment]
I've done some amount of investigation into this problem.

`lint_package()` calls `find_package()` with the argument `path="."`. This resolves to the `testthat` folder of the package. `find_package()` then iterates through the directory structure until it finds a directory with a file called `DESCRIPTION`. However, when launched from RStudio's Check function (ie. ctrl+shift+E), a copy of the package is made and `lint_package()` is started in the director `/some-path/mypackage.Rcheck/tests/testthat`. In this copy of the package, the `DESCRIPTION` file is located in `/some-path/mypackage.Rcheck/mypackage`. So, when `find_package()` iterates up the directory tree, it never finds `DESCRIPTION`.

I'm not sure what the best approach to fixing this issue would be to solve this issue.

**Comment by ndukler** [comment]
Thanks so much for this issue and the investigation, it helped me a lot. While waiting on a long term fix I think I have a workaround. I created a `setup-lint_test_file.R` file that looked like this
```
lint_path <- getwd()
```
Then when running the actual test I added the lint test like this
```
    lintr::expect_lint_free(
        path = lint_path,
        relative_path = FALSE
        )
```
This seems to work in all of the above scenarios and when running the tests with `devtools::test()`. I haven't tried it out with a continuous testing framework yet but I'm working on it and will revise this if it fails.

**Comment by ndukler** [comment]
Hmm, not sure this is working as intended.

**Comment by kloppen** [comment]
@ndukler For the benefit of anyone reading this issue thread, would you mind elaborating on the problem with your workaround?

If it helps, the hack that I've used for one of my packages is below. This problem is caused by `find_package()` returning `NULL` because it fails to correctly find the package directory when "Check" is called from the RStudio IDE. To get around this, I call `find_package()` and check that it does not return `NULL` before calling `lintr::expect_lint_free` (since `expect_lint_free` would fail in this case). However, since `find_package()` is not exported, I need to duplicate it. This definitely feels like a hack, but it is working for me until there is a more long-term solution.

For me, this is working when I run `devtools::check()`, when I run check from RStudio (since it skips linting the package), and on TravisCI.

```r
if (requireNamespace("lintr", quietly = TRUE)) {
  library(lintr)

  context("linting package")
  test_that("Package Style", {

    # A duplicate copy of the find_package function from lintr
    find_package <- function(path) {
      path <- normalizePath(path, mustWork = FALSE)

      while (!file.exists(file.path(path, "DESCRIPTION"))) {
        path <- dirname(path)
        if (identical(path, dirname(path))) {
          return(NULL)
        }
      }

      path
    }

    if (!is.null(find_package("."))) {
      expect_lint_free()
      )
    }
  })
}
```

**Comment by ndukler** [comment]
My workaround wasn't leaving me where I thought it did, resulting in everything passing but it not running the lint tests. The problem was that `R CMD check` copies everything into a weird directory structure where a bunch of directories get moved around. I've come up with a new function that works when you run just the tests and `R CMD check` from RStudio, but not with the `devtools::check()` command. It basically tries to find where the package source code with a few different lookup patterns based on how it's being run.

```
pkgname <- "tuSelecter2"
lint_path <- path.expand(normalizePath(getwd(), winslash = "/"))

find_root <- function(path, pkgname) {
  if ("DESCRIPTION" %in% dir(path)) {
    return(path.expand(normalizePath(path, winslash = "/")))
  } else if ("00_pkg_src" %in% dir(path)) {
    path <- file.path(path, "00_pkg_src", pkgname)
    return(path.expand(normalizePath(path, winslash = "/")))
  } else {
    find_root(dirname(path), pkgname)
  }
}

lint_path <- find_root(lint_path, pkgname)
```

I haven't tested it on travis yet

**Comment by kloppen** [comment]
To clarify, my workaround above skips linting when running check from the RStudio IDE, `devtools::check()` and `R CMD check`. TravisCI does still lint the package when my workaround above is in place. I have updated my comment above to clarify this.

**Comment by ndukler** [comment]
I'm probably focusing on this too much but here goes. Ok, so I figured out that it wasn't working with a command line `devtools::check()` call because it was being skipped by the `testthat::skip_on_cran()` call inside `lintr::expect_lint_free()`. It's not setting the environmental variable `NOT_CRAN` because the default behavior of `devtools::check()` is `cran=TRUE`. So that all makes sense now and everything works locally. Going to test on travis shortly.

**Comment by AshesITR** [comment]
Is this still reproducible?
I tried
```r
withr::with_envvar(c(NOT_CRAN = "true"), withr::with_dir("tests/testthat", lintr::expect_lint_free()))
```
and had no trouble.

**Comment by kloppen** [comment]
@AshesITR Yes, this is still reproducible. The steps in my first comment in this thread still produces the error, EXCEPT that you need to add `lintr` to the Suggests of the package `DESCRIPTION` in order to get it to lint the package when checking.

Please note that this this problem relates to the way that RStudio executes Check from within the IDE (ie. Ctr + Shift + E). It also now seems to affect devtools::check().

The system that I've used to verify that this is reproducible is as follows:
- Ubuntu 20.04
- RStudio 2022.02.3 Build 492
- lintr 2.01
- devtools 2.4.3

**Comment by AshesITR** [comment]
Can you update lintr to 3.0.0?

**Comment by kloppen** [comment]
@AshesITR I've updated `lintr` to 3.0.0 as requested. With this new version of `lintr` neither `devtools::check()` nor Check from the RStudio IDE produce the error previously described.

Thus, I'd consider this issue resolved. I'm closing it now, but if someone sees  a reason to keep it open, feel free to re-open.

I'd like to thank the entire `lintr` team for resolving this and for all the great work that you've done on this package.

**Comment by MichaelChirico** [comment]
Thanks for the report! Glad it's resolved.
